Bluetooth: sample: Add central health thermometer and peripheral sensor integration

charanpv opened the pull request:
Integrated temperature sensor sample fetch and use of the value
in the GATT indications.

Added central sample to read die temperature value from peripheral
device

Signed-off-by: Charan Pudiyaneravana Venkatesh <charan.pv@sixoctets.com>

charanpv committed:
samples: Bluetooth: Integrate temperature sensor sample fetch

Integrated temperature sensor sample fetch and use of the value
in the GATT indications.

Signed-off-by: Charan Pudiyaneravana Venkatesh <charan.pv@sixoctets.com>
Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

charanpv committed:
samples: Bluetooth: Added central health thermometer

Added central sample to read die temperature value from peripheral
device.

Signed-off-by: Charan Pudiyaneravana Venkatesh <charan.pv@sixoctets.com>
Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented on samples/bluetooth/peripheral_ht/src/hts.c:
```diff
@@ -53,26 +56,65 @@ BT_GATT_SERVICE_DEFINE(hts_svc,
+
 	if (simulate_htm) {
+		static u8_t htm[5];
+		double temperature;
```
temperature needs to be static variable for simulated values. And assign to 20U.

charanpv replied:
assigned 20U value for temperature and its working fine

joerchan commented on samples/bluetooth/central_ht/src/main.c:
```diff
@@ -0,0 +1,277 @@
+	 */
+	struct bt_le_scan_param scan_param = {
+		.type       = BT_HCI_LE_SCAN_ACTIVE,
+		.filter_dup = BT_HCI_LE_SCAN_FILTER_DUP_DISABLE,
```
Don't use BT_HCI defines, use 0 instead here. 

Is there any reason why you set Filter Duplicates to disabled?
Otherwise BT_LE_SCAN_ACTIVE could be used instead for this.

charanpv replied:
well, I have changed it to BT_LE_SCAN_ACTIVE now 
thank you...

joerchan commented on samples/bluetooth/central_ht/src/main.c:
```diff
@@ -0,0 +1,277 @@
+		printk("[UNSUBSCRIBED]\n");
+		params->value_handle = 0U;
+		return BT_GATT_ITER_STOP;
+	}
```
Blank line after `}`

charanpv replied:
Sure,  I will add blank line

joerchan commented on samples/bluetooth/central_ht/src/main.c:
```diff
@@ -0,0 +1,277 @@
+		params->value_handle = 0U;
+		return BT_GATT_ITER_STOP;
+	}
+	ieee = (u32_t *)&((u8_t *)data)[1];
```
I think we should create helper functions for this. Could you add that?
bt_float convert functions would be nice.

jhedberg replied:
Additionally this looks like unaligned access, so it should be using `UNALIGNED_GET()` as far as I see

joerchan replied:
Might be good to enable unaligned fault to make sure that it is done correctly.
```
SCB->CCR |= SCB_CCR_UNALIGN_TRP_Msk;
```

charanpv replied:
I used sys_pt_le24 function and it is working fine

charanpv replied:
I used sys _put and sys_get function for unaligned access.

jhedberg commented on samples/bluetooth/peripheral_ht/src/hts.c:
```diff
@@ -53,26 +56,64 @@ BT_GATT_SERVICE_DEFINE(hts_svc,
+
+		htm[0] = 0; /* temperature in celcius */
+		ieee = (u32_t *)&htm[1];
+		*ieee = ((u32_t)exponent << 24) | mantissa;
```
How is this alignment-safe? Seems like you should be using `UNALIGNED_PUT()`

charanpv replied:
can i use sys_put function for unaligned memory location ?

joerchan replied:
Asking the right question! :)

Yes, you should use sys_put_le and sys_get_le, because the HTS specification states that little endian is the byte transmission order. There is also sys_put_le24 which would come in handy in this case.

You can check that it works with the nRF Connect for Mobile App.

charanpv replied:
As u said , i used sys function and it works with nRF connect as well as with central_ht

jhedberg reviewed (changes requested):
See my two inline comments

joerchan commented on samples/bluetooth/central_ht/src/main.c:
```diff
@@ -0,0 +1,275 @@
+static struct bt_gatt_discover_params discover_params;
+static struct bt_gatt_subscribe_params subscribe_params;
+
+static double power(double x, s8_t y)
```
math.h pow function take double, double, should we perhaps put this into minimal-libc in zephyr with that prototype?

cvinayak replied:
there is reference to pow in drivers/sensor/light_sensor, from where (library) does this get referenced? If pow already exists/used we should reuse that implementation.

joerchan replied:
@cvinayak It comes from math.h, but libc-minimal zephyrs-builtin does not have it, if using newlibc then it is available:
That is why it looks like this:
```
#if defined(CONFIG_NEWLIB_LIBC)
#include <math.h>

#define GPTP_POW2(exp) pow(2, exp)
```

cvinayak commented on samples/bluetooth/peripheral_ht/src/hts.c:
```diff
@@ -53,26 +56,64 @@ BT_GATT_SERVICE_DEFINE(hts_svc,
+		exponent = (u8_t)-2;
+
+		htm[0] = 0; /* temperature in celcius */
+		sys_put_le24(mantissa, (u8_t*)&htm[1]);
```
missing space in between u8_t*

charanpv replied:
Added space between u8_t* 
Thank you

joerchan commented:
@jhedberg Merge this one?

jhedberg commented:
@joerchan thanks for the reminder. Will merge once the shippable rerun completes. It’s unsafe to merge PRs whose shippable ran weeks (or even months in this case) ago.

jhedberg commented:
@joerchan if you look at the Shippable result you'll see I was justified with my paranoia :)

@charanpv could you update the PR so that the sample doesn't use these deprecated APIs anymore?

joerchan commented:
> @joerchan if you look at the Shippable result you'll see I was justified with my paranoia :)

Just because you're paranoid doesn't mean they aren't after you. :)

cvinayak commented:
@jhedberg @joerchan please re-review, I have rebased and ported to new APIs.

joerchan commented:
> @jhedberg @joerchan please re-review, I have rebased and ported to new APIs.

@cvinayak After the rebase `BT_LE_CONN_OPT_NONE` must be renamed to `BT_CONN_LE_OPT_NONE`, or `BT_CONN_LE_CREATE_CONN` could be used directly in the call.

joerchan commented on samples/bluetooth/central_ht/src/main.c:
```diff
@@ -0,0 +1,280 @@
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	if (conn_err) {
+		printk("Failed to connect to %s (%u)\n", addr, conn_err);
```
Should we go back to scan here?
In that case default_conn needs to be unref'ed and set to NULL as well.

cvinayak replied:
Yes, thats is something I do even for other central samples. I will try sending an additional commit for this.

joerchan reviewed (approved):
Looks good. Just a question about how to handle connected with err. Now the sample needs to be reset if that happens.

cvinayak committed:
samples: Bluetooth: Restart scanning on connection failure

Restart scanning if connection creation fails; could happen
when devices with low RSSI are discovered but not available
thereafter to establish a connection.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
> how to handle connected with err. Now the sample needs to be reset if that happens.

As you already suggested, I have added an additional commit to restart scanning.

jhedberg commented on samples/bluetooth/central/src/main.c:
```diff
@@ -53,6 +55,10 @@ static void device_found(const bt_addr_le_t *addr, s8_t rssi, u8_t type,
 				BT_LE_CONN_PARAM_DEFAULT, &default_conn);
 	if (err) {
 		printk("Create conn to %s failed (%u)\n", addr_str, err);
+
+		default_conn = NULL;
```
What's going on here? If `bt_conn_le_create()` returns failure it should leave the conn pointer untouched. If that's not happening we shouldn't be working around it on the application side like this but fix the actual issue.

joerchan replied:
conn is only assigned when returning success, so this line is not needed.

jhedberg commented on samples/bluetooth/central_ht/src/main.c:
```diff
@@ -0,0 +1,291 @@
+				continue;
+			}
+
+			err = bt_conn_le_create(addr, BT_CONN_LE_CREATE_CONN,
+						BT_LE_CONN_PARAM_DEFAULT,
+						&default_conn);
+			if (err) {
+				printk("Create connection failed (err %d)\n",
+				       err);
+
+				default_conn = NULL;
```
Same thing here. As @joerchan said this line is not needed.

jhedberg commented on samples/bluetooth/central_hr/src/main.c:
```diff
@@ -132,6 +134,10 @@ static bool eir_found(struct bt_data *data, void *user_data)
 			if (err) {
 				printk("Create conn failed (err %d)\n", err);
+
+				default_conn = NULL;
```
And here

jhedberg commented:
Could you also enable this in CI? Seems like nothing was build tested:

<img width="676" alt="Screenshot 2020-05-06 at 13 01 46" src="https://user-images.githubusercontent.com/1032570/81164914-c076c980-8f99-11ea-8026-777b6391f3d3.png">

jhedberg commented:
Or actually, does the above mean that only these samples were built, and that's why the build was so fast?

jhedberg commented:
Answering myself: yes, CI seems to be nicely optimised now to only build the relevant tests.

joerchan commented:
@jhedberg Look like CI ran a shorter build.

```
INFO    - Adding tasks to the queue...
INFO    - 1/2 qemu_cortex_m3            samples/bluetooth/peripheral_ht/sample.bluetooth.peripheral_ht PASSED (build)
Renaming output directory to /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/sanity-out.1
INFO    - 2/2 qemu_x86                  samples/bluetooth/peripheral_ht/sample.bluetooth.peripheral_ht PASSED (build)
INFO    - 2 of 2 tests passed (100.00%), 0 failed, 0 skipped with 0 warnings in 13.99 seconds
```
